### PR TITLE
Make left zero dynamic (rewrite)

### PR DESCRIPTION
**Pull request by cannorin**
Related: #501

### CODE REVIEW

**Comment by cannorin** [comment]
perf script:
```fsharp
#r "src/FSharpPlus/bin/Release/netstandard2.0/FSharpPlus.dll"
// #r "nuget:FSharpPlus"
open System
open FSharpPlus

// try `traverse`
let xs = traverse Some [1;2;3]
let ys = traverse List.singleton [1;2;3]

// type without IsLeftZero or Empty
type MyWrapped<'a> = Wrapped of 'a with
  static member Return (x: 'a) = Wrapped x
  static member (>>=) (Wrapped a, f) = f a

// type with Empty
type Maybe<'a> = Just of 'a | Nothing with
  static member Empty : Maybe<'a> = Nothing

// type with IsLeftZero
type MyList<'a> = MyCons of 'a * MyList<'a> | MyNil with
  static member IsLeftZero(x: MyList<'a>) =
    match x with MyNil -> true | _ -> false
  member this.ToSeq() : 'a seq =
    seq {
      match this with
      | MyNil -> ()
      | MyCons (a, next) ->
        yield a
        yield! next.ToSeq()
    }
  interface Collections.Generic.IEnumerable<'a> with
    member this.GetEnumerator() = this.ToSeq().GetEnumerator()
  interface Collections.IEnumerable with
    member this.GetEnumerator() = this.ToSeq().GetEnumerator() :> _

// perf
open FSharpPlus.Control

let [<Literal>] N = 1000000

let before = DateTime.Now
for _ = 1 to N do
  assert not (IsLeftZero.Invoke(Some 42))
  assert IsLeftZero.Invoke(None : int option)
  assert not (IsLeftZero.Invoke([4; 2]))
  assert IsLeftZero.Invoke([] : int list)
  assert not (IsLeftZero.Invoke([|"4"; "2"|]))
  assert IsLeftZero.Invoke([||] : string array)
  assert not (IsLeftZero.Invoke(Seq.singleton 42))
  assert IsLeftZero.Invoke(Seq.empty : int seq)
  assert not (IsLeftZero.Invoke(Ok 42 : Result<int, string>))
  assert IsLeftZero.Invoke(Error "msg" : Result<int, string>)
  assert not (IsLeftZero.Invoke(Wrapped 42))
  assert not (IsLeftZero.Invoke(Choice1Of2 42 : Choice<int, string>))
  assert IsLeftZero.Invoke(Choice2Of2 "msg" : Choice<int, string>)
  assert not (IsLeftZero.Invoke(Just 42))
  assert IsLeftZero.Invoke(Nothing : Maybe<int>)
  assert not (IsLeftZero.Invoke(MyCons (4, MyCons (2, MyNil))))
  assert IsLeftZero.Invoke(MyNil : MyList<int>)
let after = DateTime.Now
let total = (after - before).TotalMilliseconds

printfn "OK: %fms total, %fms avg" (total) (total / float N)
```

The result on my PC: 9ms to 12ms of overhead regardless of:
* The number of `N`
* The number of `IsLeftZero.Invoke` calls
* The number of types used

File: src/FSharpPlus/Control/Applicative.fs [hunk 1]
```diff
@@ -143,21 +143,92 @@ type Lift3 with
...
+    static member Option (x: option<'t>) = Option.isNone x
+    static member Result (x: Result<'t, 'u>) = match x with Error _ -> true | _ -> false
+    static member Choice (x: Choice<'t, 'u>) = match x with Choice2Of2 _ -> true | _ -> false
+
+#if !FABLE_COMPILER
+type IsLeftZeroHelper<'a>() =
+    /// turns false if
+    /// - it should always return false because neither `IsLeftZero` nor `Empty` are present
+    /// - the target method is inlined and cannot be called through reflection
+    static let mutable isValid = true
```
Comment (wallymathieu) [inline src/FSharpPlus/Control/Applicative.fs:160]: That's an interesting trick 😃

Comment (cannorin) [reply src/FSharpPlus/Control/Applicative.fs:160]: yeah, it seems the compiler does not add any kind of metadata to the uncallable inline methods 😅

File: src/FSharpPlus/Control/Applicative.fs [hunk 2]
```diff
@@ -143,21 +143,92 @@ type Lift3 with
...
+            let meth = helperTy.GetMethod(helperName).MakeGenericMethod(tprms)
+            fun x -> meth.Invoke(null, [|box x|]) |> unbox
+        let targs = ty.GetGenericArguments()
+        if check typedefof<seq<_>> then helper "Seq" targs
+        else if check typedefof<NonEmptySeq<_>> then helper "NonEmptySeq" targs
+        else if check typedefof<list<_>> then helper "List" targs
+        else if ty.IsArray then helper "Array" [| ty.GetElementType() |]
+        else if check typedefof<option<_>> then helper "Option" targs
+        else if check typedefof<Result<_, _>> then helper "Result" targs
+        else if check typedefof<Choice<_, _>> then helper "Choice" targs
```
Comment (gusty) [inline src/FSharpPlus/Control/Applicative.fs:176]: Just nitpicking, can we use `elif` instead of `else if` :)

**Review by gusty — APPROVED** [review]

**Pull request merged**